### PR TITLE
enhance shape inference, fix bugs

### PR DESCRIPTION
**Pull request by zhijxu-MS**
1 add shape inference function of gather/unpack/concat/minimum

2  fix several bugs found when running real models

### CODE REVIEW

File: tf2onnx/onnx_opset/nn.py [hunk 1]
```diff
@@ -150,7 +150,8 @@ def add_padding(ctx, node, kernel_shape, strides, dilations=None, spatial=2):
                 input_shape = spatial_map(input_shape, constants.NHWC_TO_NCHW)
                 output_shape = spatial_map(output_shape, constants.NHWC_TO_NCHW)
             # calculate pads
-            if any(input_shape[i + 2] == -1 or output_shape[i + 2] == -1 for i in range(spatial)):
+            if any(input_shape[i + 2] == -1 or output_shape[i + 2] == -1 for i in range(spatial)) \
+               or any(output_shape[i + 2] == -1 for i in range(spatial)):
```
Comment (nbcsm) [inline tf2onnx/onnx_opset/nn.py]: this seems already fixed?
```
any(input_shape[i + 2] == -1 or output_shape[i + 2] == -1 for i in range(spatial)) 
```

File: tf2onnx/shape_inference.py [hunk 2]
```diff
@@ -213,6 +212,36 @@ def infer_input_shapes(g, node):
 
 
 def infer_output_shapes_with_partial_inputs(g, node):
+    # output shape of concat op: only the dim val of concatenated dim will be changed
+    # so only partial(at least one) input shapes need to be known to infer output shape of concat node
+    if rnn_utils.is_concat_op(node):
```
Comment (nbcsm) [inline tf2onnx/shape_inference.py]: how about move is_xxx_node to utils?

File: tf2onnx/onnx_opset/tensor.py [hunk 3]
```diff
@@ -532,11 +532,12 @@ def version_4(cls, ctx, node, **kwargs):
...
                 raise ValueError("StridedSlice: attribute " + attr_name + " not supported")
-        input_shape = ctx.get_shape(node.input[0])
-        begin = node.inputs[1].get_tensor_value(as_list=False)
-        end = node.inputs[2].get_tensor_value(as_list=False)
-        strides = node.inputs[3].get_tensor_value(as_list=False)
-        max_size = np.iinfo(begin.dtype).max
+        onnx_dtype = ctx.get_dtype(node.input[1])
+        np_dtype = utils.ONNX_TO_NUMPY_DTYPE[onnx_dtype]
+        max_size = np.iinfo(np_dtype).max
+        begin = node.inputs[1].get_tensor_value()
```
Comment (nbcsm) [inline tf2onnx/onnx_opset/tensor.py]: tbd

File: tf2onnx/shape_inference.py [hunk 4]
```diff
@@ -115,43 +116,18 @@ def infer_shape_for_node(g, node):
...
-        logger.debug("set ConcatV2 node [%s] with new shape %s", node.output[0], new_shape)
-        return True
-
     if node.type == "Gather":
         # uses the follwing link to know how to infer shape of output
         # https://www.tensorflow.org/api_docs/python/tf/gather
         shape_params = g.get_shape(node.input[0])
         shape_indices = g.get_shape(node.input[1])
-        axis = node.input[2].get_tensor_value()
+        # in lower tf version, gather only has 2 inputs
```
Comment (nbcsm) [inline tf2onnx/shape_inference.py]: which version? please add to comment.

Comment (zhijxu-MS) [reply tf2onnx/shape_inference.py]: https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/gather.html

Comment (zhijxu-MS) [reply tf2onnx/shape_inference.py]: will add related link to the comment

**Pull request merged**